Implement MongoDB datasource and FDAs (#192)
telefonicaid/fiware-data-access: doc/03_api.md

| `timeColumn`                                             | ✓        | string        | Required with `refreshPolicy` of type `window` and `partition`. Column in the table indicating when the data was received (date).                                                          |
| `cached`                                                 | ✓        | boolean       | If `false`, the FDA is created as only-fresh: no parquet snapshot is maintained, no DAs are allowed, and the FDA is queried through `GET /{visibility}/fdas/{fdaId}/data`. Default `true`. |
| `datasourceId`                                           | ✓        | string        | Datasource id used to resolve DB credentials for this FDA. If omitted, FDA uses `default`.                                                                                                 |
| `collection`                                             | ✓ (\*)   | string        | Mongo source collection name. Mandatory when datasource type is `mongodb`.                                                                                                                 |

> | `attrs`                                                  | ✓ (\*)   | array[string] | Ordered list of Mongo fields to project and persist. Must be a non-empty array when datasource type is `mongodb`.                                                                          |

Reviewer: As an alternative, we could use the `query` field for this

```
"query": "db.<collection>.find({<query>},{<projection>})"
```

It's ugly... but having two conditional fields like these is also ugly :)

It's more homogenous with the PG case, as in the PG case the table (equivalent to collection) and projecting are included in the SELECT statement used in `query`.

Reviewer: It is also more powerfull, as not only find() queries could be included, also agregation ones, eg:

```
"query": "db.<collection>.aggregate([{$group: {_id: "$category", "n": {$sum: 1}}}])"
```

Author: @fgalan 
I agree that collection/projection conceptually belong to the query definition.

However, instead of accepting Mongo shell strings (`db.collection.find(...)`), I'd suggest keeping a structured JSON representation:

```json
{
  "query": {
    "collection": "sensors",
    "filter": {...},
    "projection": [...]
  }
}
```

This keeps validation simple, avoids parsing Mongo shell syntax and leaves room for future extensions (aggregation pipelines, etc.).

Regarding the shell-string approach, my main concern is not security (FDA creation should be secured), but maintainability and API consistency:

* We'd need to parse the query string to extract information such as collection, projection or operation type. With a structured object these elements are already explicit. (There is no way to execute query strings in Mongo library without json parsing).
* Validation becomes much simpler. For example, checking that a collection exists, that a projection is present, or that only supported operations are used.
* It makes future evolution easier. We could add support for aggregation pipelines or other Mongo-specific features without having to build and maintain a Mongo query parser.
* It keeps the API self-describing. Consumers can easily understand the query structure without knowing Mongo shell syntax.

That said, if we explicitly want to support arbitrary Mongo operations from day one (find, aggregate, etc.), the string approach is certainly more flexible and closer to how SQL queries are handled today.

Reviewer: Indeed this is an important diference... the PG API allows to exceute a "blind" text string (as SQL query) while MongoDB API doesn't. Thus, using an structured `query` param in this case makes sense.

So we can go with that idea with an addition: make the field "polymorphic" to support different query families in MongoDB. In fact, as far as I remember/know at the present moment there are only two families: the conventional query (i.e. `find()`) and the aggregation pipeline, so we could have:

* `query` as string: only for PG (to hold a SQL query), not to be used with MongoDB datasources
* `query` for find with the following fields:

```
{
  "query": {
    "collection": "sensors",
    "filter": {...},
    "projection": [...]
  }
}
```

* `query` for aggregation with the following fields:

```
{
  "query": {
    "collection": "sensors",
    "aggregation": <agregation pipeline>
  }
}
```

Eg. of the third type:

```
{
  "query": {
    "collection": "sensors",
    "aggregation": [
      { "$group": {"_id": "$category", "n": {"$sum": 1}}}
    ]
  }
}
```

Reviewer: Some additional comments:

* To be closer to the MongoDB API, better to use `projection` as object, e.g:

```
{_id:0, credits: 1}
```

* Some stages in the aggregation pipeline has effect beyond querying. For instance, the [$out stage](https://www.mongodb.com/docs/manual/reference/operator/aggregation/out/). Maybe we should filter out sensitive stages like that (although for the first version of this functionality we can bypass that problem, just mentioning it in the documentation)

Author: I like this approach.

Using a polymorphic `query` field keeps the API consistent while avoiding the need to parse Mongo shell syntax.

I also agree that using MongoDB's native projection format is preferable to an array of attributes.

One thing we should define is whether the two Mongo query variants are mutually exclusive. My preference would be:

* `collection` mandatory in both cases.
* Either `filter` (optionally with `projection`) or `aggregation`.
* Never both at the same time.

Regarding aggregation stages such as `$out`, I agree, we can initially document that aggregation pipelines are not restricted and revisit that later if needed. Since FDA creation is already a privileged operation, I don't see it as a blocker for the first version.

Author: Done

| [`objStgConf`](#object-storage-configuration-objstgconf) | ✓        | object        | Various options to configure the FDA uploaded in the object storage app.                                                                                                                   |
| `timeColumn`                                             | ✓        | string        | Required with `refreshPolicy` of type `window` and `partition`. Column in the table indicating when the data was received (date).                                                          |
| `cached`                                                 | ✓        | boolean       | If `false`, the FDA is created as only-fresh: no parquet snapshot is maintained, no DAs are allowed, and the FDA is queried through `GET /{visibility}/fdas/{fdaId}/data`. Default `true`. |
| `datasourceId`                                           | ✓        | string        | Datasource id used to resolve DB credentials for this FDA. If omitted, FDA uses `default`.                                                                                                 |

> (\*) Mandatory only for Mongo datasource FDAs.

Reviewer: No longer used?

Author: Yes, fixed in https://github.com/telefonicaid/fiware-data-access/pull/192/commits/762ef672ac907a2cd84bc89b8bd837660be42a8e

For MongoDB datasources, `query` contains:

| Field        | Type   | Description                                                                |
| ------------ | ------ | -------------------------------------------------------------------------- |
| `collection` | string | MongoDB collection name.                                                   |
| `filter`     | object | MongoDB filter document.                                                   |
| `projection` | object | MongoDB projection document defining the fields materialized into the FDA. |

> Nested MongoDB fields can be projected using dot notation:

Reviewer: In fact, projection in MongoDB could be more complex than just 0 and 1. See gory details at https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results ;)

I'd suggest to cite that document

Author: Added in https://github.com/telefonicaid/fiware-data-access/pull/192/commits/762ef672ac907a2cd84bc89b8bd837660be42a8e

```json
{
    "projection": {
        "device.name": 1,
        "status": 1
    }
}
```

Projected nested fields are materialized as FDA columns preserving their dot notation (`device.name`).

When generating defaultDataAccess parameters, dots are replaced by underscores. For example:

-   Column: `device.name`
-   Parameter: `device_name`

Datasource-specific constraints:

-   Mongo datasource FDAs are currently cached-only (`cached=true`).
-   Mongo datasource FDAs do not support `refreshPolicy.type=window`.

> -   If `timeColumn` is provided for Mongo FDAs, it must be included in `query.projection`.

Author: This section may be moved to a dedicated Advanced Topics file covering MongoDB-specific queries in the future.

> -   If `timeColumn` is provided for Mongo FDAs, it must be included in `query.projection`.

Reviewer: ```suggestion
-   If `timeColumn` is provided for Mongo FDAs, it must be included in `query.projection` (or the `query.projection` omitted at all as in this case no projection is done and all fields are retrieved).
```

Is that correct?

Author: Yes, that's correct

A FDA is represented by a JSON object with the following fields:

| Parameter                                                | Optional | Type          | Description                                                                                                                                                                                |
| -------------------------------------------------------- | -------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `id`                                                     |          | string        | FDA unique identifier                                                                                                                                                                      |
| `description`                                            | ✓        | string        | A free text used by the client to describe the FDA. If omitted, no description is stored.                                                                                                  |

> | `query`                                                  |          | string/object | Source query definition. For `postgres`, a SQL query string. For `mongodb`, a Mongo query definition containing `collection`, `filter` and `projection`.                                   |

Reviewer: ```suggestion
| `query`                                                  |          | string/object | Source query definition. For `postgres`, a SQL query string. For `mongodb`, a Mongo query definition detailed below.                                   |
```

So when we improve this (to implement https://github.com/telefonicaid/fiware-data-access/issues/197) this row doesn't need to be changed.

| [`objStgConf`](#object-storage-configuration-objstgconf) | ✓        | object        | Various options to configure the FDA uploaded in the object storage app.                                                                                                                   |
| `timeColumn`                                             | ✓        | string        | Required with `refreshPolicy` of type `window` and `partition`. Column in the table indicating when the data was received (date).                                                          |
| `cached`                                                 | ✓        | boolean       | If `false`, the FDA is created as only-fresh: no parquet snapshot is maintained, no DAs are allowed, and the FDA is queried through `GET /{visibility}/fdas/{fdaId}/data`. Default `true`. |
| `datasourceId`                                           | ✓        | string        | Datasource id used to resolve DB credentials for this FDA. If omitted, FDA uses `default`.                                                                                                 |

> For MongoDB datasources, `query` contains:

Reviewer: ```suggestion
For MongoDB datasources, `query` is an object with the following keys:
```

| Field        | Type   | Description                                                                |
| ------------ | ------ | -------------------------------------------------------------------------- |
| `collection` | string | MongoDB collection name.                                                   |
| `filter`     | object | MongoDB filter document.                                                   |
| `projection` | object | MongoDB projection document defining the fields materialized into the FDA. |

In MongoDB, projection can include more complex operators like `$slice` or `$elemMatch`. See the
[MongoDB projection documentation](https://www.mongodb.com/docs/manual/tutorial/project-fields-from-query-results/) for
details. Nested MongoDB fields can be projected using dot notation:

```json
{
    "projection": {
        "device.name": 1,
        "status": 1
    }
}
```

Projected nested fields are materialized as FDA columns preserving their dot notation (`device.name`).

> When generating defaultDataAccess parameters, dots are replaced by underscores. For example:

Reviewer: Maybe a link the doc explaining what the defaultDataAccess is should be included here.

Author: Added link in https://github.com/telefonicaid/fiware-data-access/pull/192/commits/94a967f4d21e3081f6572b45988a429bff0a8285
